File: python/maximum_product_subarray.py
class Solution:
    """
    @param nums: An array of integers
    @return: An integer
    """
    def maxProduct(self, nums):
        # write your code here
        if not nums:
            return 0
        mini, maxi, res = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            if nums[i]>0 :
                maxi = max(maxi*nums[i], nums[i])
                mini = min(mini*nums[i], nums[i])
            else:
                tmp = maxi
                maxi = max(mini*nums[i], nums[i])
                mini = min(tmp*nums[i], nums[i])
            res = max(res, maxi)
        return res

test = Solution()
nums = [2,3,-1,4]
res = test.maxProduct(nums)
print(res)

File: cpp/favourite/MaximumProductSubarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        if(!n) return 0;
        int minNum = nums[0], maxNum=nums[0];
        int res = maxNum;
        for(int i=1; i<nums.size(); i++) {
            int preMin = minNum, preMax = maxNum;
            maxNum = max(nums[i], max(preMin*nums[i], preMax*nums[i]));
            minNum = min(nums[i], min(preMin*nums[i], preMax*nums[i]));
            res = max(maxNum, res);
        }
        return res;
        
    }
};